cs/src/core/Device/AzurePageBlobDevice.cs

```diff
@@ -0,0 +1,136 @@
+        {
+            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(connectionString);
+            CloudBlobClient client = storageAccount.CreateCloudBlobClient();
+            container = client.GetContainerReference(containerName);
+            container.CreateIfNotExists();
+            blobs = new ConcurrentDictionary<int, CloudPageBlob>();
+            this.blobName = blobName;
+            this.deleteOnClose = deleteOnClose;
+        }
+
+        /// <summary>
+        /// <see cref="IDevice.Close">Inherited</see>
+        /// </summary>
+        public override void Close()
+        {
+            // Unlike in LocalStorageDevice, we explicitly remove all page blobs if the deleteOnClose flag is set, instead of relying on the operating system
+            // to delete files after the end of our process. This leads to potential problems if multiple instances are sharing the same underlying page blobs.
+            //
+            // Since this flag is presumably only used for testing though, it is probably fine.
+            if (deleteOnClose)
+            {
+                foreach (var entry in blobs)
+                {
+                    entry.Value.Delete();
+                }
+            }
+        }
+        /// <summary>
+        /// <see cref="IDevice.Close">Inherited</see>
+        /// </summary>
+        public override void DeleteSegmentRange(int fromSegment, int toSegment)
+        {
+            for (int i = fromSegment; i < toSegment; i++)
+            {
+                if (blobs.TryRemove(i, out CloudPageBlob blob))
+                {
+                    blob.Delete();
+                }
+            }
+        }
+        /// <summary>
+        /// <see cref="IDevice.ReadAsync(int, ulong, IntPtr, uint, IOCompletionCallback, IAsyncResult)">Inherited</see>
+        /// </summary>
+        public override unsafe void ReadAsync(int segmentId, ulong sourceAddress, IntPtr destinationAddress, uint readLength, IOCompletionCallback callback, IAsyncResult asyncResult)
+        {
+            CloudPageBlob pageBlob = GetOrAddPageBlob(segmentId);
+
+            // Even though Azure Page Blob does not make use of Overlapped, we populate one to conform to the callback API
+            Overlapped ov = new Overlapped(0, 0, IntPtr.Zero, asyncResult);
+            NativeOverlapped* ovNative = ov.UnsafePack(callback, IntPtr.Zero);
+
+            UnmanagedMemoryStream stream = new UnmanagedMemoryStream((byte*)destinationAddress, readLength, readLength, FileAccess.Write);
+
+            // TODO(Tianyu): This implementation seems to swallow exceptions that would otherwise be thrown from the synchronous version of this
+            // function. I wasn't able to find any good documentaiton on how exceptions are propagated or handled in this scenario. 
+            pageBlob.BeginDownloadRangeToStream(stream, (Int64)sourceAddress, readLength, ar => callback(0, readLength, ovNative), asyncResult);
```

Comment:
>ar => callback(0, readLength, ovNative) [](start = 90, length = 39)

What are the guarantees of this lambda? We need to ensure that the callback is invoked exactly once, and if it's not invoked at all, free the overlapped, and invoke callback with the appropriate error.